[SPARK-4835] Disable validateOutputSpecs for Spark Streaming jobs

JoshRosen opened the pull request:
This patch disables output spec. validation for jobs launched through Spark Streaming, since this interferes with checkpoint recovery.

Hadoop OutputFormats have a `checkOutputSpecs` method which performs certain checks prior to writing output, such as checking whether the output directory already exists.  SPARK-1100 added checks for FileOutputFormat, SPARK-1677 (#947) added a SparkConf configuration to disable these checks, and SPARK-2309 (#1088) extended these checks to run for all OutputFormats, not just FileOutputFormat.

In Spark Streaming, we might have to re-process a batch during checkpoint recovery, so `save` actions may be called multiple times.  In addition to `DStream`'s own save actions, users might use `transform` or `foreachRDD` and call the `RDD` and `PairRDD` save actions.  When output spec. validation is enabled, the second calls to these actions will fail due to existing output.

This patch automatically disables output spec. validation for jobs submitted by the Spark Streaming scheduler.  This is done by using Scala's `DynamicVariable` to propagate the bypass setting without having to mutate SparkConf or introduce a global variable.

JoshRosen committed:
[SPARK-4835] Disable validateOutputSpecs for Spark Streaming jobs.

tdas commented:
So basically you want to split this PR into two - (1) fix the bug, and (2) introduce an override. I like the idea. Lets do that.

JoshRosen committed:
Deduplicate isOutputSpecValidationEnabled logic.

JoshRosen committed:
Revise disableOutputSpecValidation() comment to not refer to Spark Streaming.

JoshRosen committed:
Remove Streaming-specific setting to undo this change; update conf. guide

JoshRosen committed:
Add test case in Streaming; fix bug for transform()

SparkQA commented:
[Test build #25003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25003/consoleFull) for   PR 3832 at commit [`6485cf8`](https://github.com/apache/spark/commit/6485cf880465cf7bd8e501dc861869be58029995).
 * This patch merges cleanly.

JoshRosen commented:
@tdas I've updated this PR and added a test case.  My test case uses calls inside of a `transform()` call to emulate what Streaming's `saveAsHadoopFiles` operation does.  Is this a valid use of `transform()` or am I breaking rules by having actions in my transform function?  My gut says that we shouldn't endorse / recommend this for the same reason that we advise against using accumulators inside of map() tasks: the transform call might get evaluated multiple times if caching isn't use, which makes it possible to write programs whose behavior changes depending on whether caching is enabled.

I wasn't able to get the existing "recovery with saveAsNewAPIHadoopFiles operation" test to fail, though, even though I discovered this bug while refactoring that test in my other PR.  I think that the issue is that the failed `saveAsNewAPIHadoopFiles` jobs failed but did not trigger a failure of the other actions / transformations in that batch, so we still got the correct output even though the batch completion event wasn't posted to the listener bus.  The current tests rely on wall-clock time to detect when batches have been processed and hence didn't detect that the batch completion event was missing.  I noticed that the StreamingListener API doesn't really have any events for job / batch failures, but that's a topic for a separate PR.

I was about to write that this bug might not actually affect users who don't use `transform` but it still impacts users in the partial-failure case where they've used PairDStreamFunctions.saveAsHadoopFiles() but a batch fails with partially-written output: an individual output _partition_ might be atomically committed to the output directory (e.g. if the file exists, then it has the right contents), but I think we can still wind up in a scenario where only a subset of the partitions are written and the non-empty output directory prevents the recovery from recomputing the missing partitions.

SparkQA commented:
[Test build #25003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25003/consoleFull) for   PR 3832 at commit [`6485cf8`](https://github.com/apache/spark/commit/6485cf880465cf7bd8e501dc861869be58029995).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25003/
Test PASSed.

tdas commented:
Yes, this is hardly a valid used case of `transform`. I took a look at the unit test, and I think I know why it works. But this unit test is very very obfuscated, and hard to understand why it works. Does it even fail reliably without the fix? Because if it does not fail reliably without the fix then there isnt much point of having it. If it fails reliably without the fix, and does not fail reliably with the fix, then I am okay with unit test as long as there is sufficient comments explaining how this unit test works. Maybe we can open another JIRA mentioning the exceptions that do not get propagated which makes this unit test so complicated. And refer to the JIRA in the comments around the unit test. Other than that I think I am okay. For this subtle an issue (validity checks), its better to have some unit test than have none at all.

JoshRosen commented:
The new test added here always failed before the fix in this patch.  I agree that it's a bit convoluted; it would be better to catch this bug in the original `saveAs*` tests, but those tests won't be able to catch it until they're rewritten to use StreamingTestWaiter.  I can add a comment / TODO to remove this test once we've strengthened the other tests in a later PR.

JoshRosen committed:
Add comment explaining use of transform() in test.

JoshRosen commented:
Alright, I've pushed one more commit that explains the confusing `transform()` use and refers to https://issues.apache.org/jira/browse/SPARK-5079.  Feel free to fix up / adjust the wording yourself on commit.

SparkQA commented:
[Test build #25040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25040/consoleFull) for   PR 3832 at commit [`36eaf35`](https://github.com/apache/spark/commit/36eaf355a17a9e6568d33257a84b4ad35e72b67f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25040/consoleFull) for   PR 3832 at commit [`36eaf35`](https://github.com/apache/spark/commit/36eaf355a17a9e6568d33257a84b4ad35e72b67f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25040/
Test PASSed.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -255,6 +255,45 @@ class CheckpointSuite extends TestSuiteBase {
+    // Regression test for SPARK-4835.
+    //
+    // In that issue, the problem was that `saveAsHadoopFile(s)` would fail when the last batch
+    // was restarted from a checkpoint since the output directory would already exist.  However,
```
nit: extra space before `However`

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -255,6 +255,45 @@ class CheckpointSuite extends TestSuiteBase {
+    // was restarted from a checkpoint since the output directory would already exist.  However,
+    // the other saveAsHadoopFile* tests couldn't catch this because they only tested whether the
+    // output matched correctly and not whether the post-restart batch had successfully finished
+    // without throwing any errors.  The following test reproduces the same bug with a test that
```
nit: extra space before `The following`

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -255,6 +255,45 @@ class CheckpointSuite extends TestSuiteBase {
+            val output = rdd.map(x => (x, 1)).reduceByKey(_ + _)
+            output.saveAsHadoopFile(
+              new File(tempDir, "result-" + time.milliseconds).getAbsolutePath,
+              classOf[Text],
```
Scala style, isnt the indenting wrong here?

tdas replied:
No it isnt, sorry.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -255,6 +255,45 @@ class CheckpointSuite extends TestSuiteBase {
+              new File(tempDir, "result-" + time.milliseconds).getAbsolutePath,
+              classOf[Text],
+              classOf[IntWritable],
+              classOf[TextOutputFormat[Text, IntWritable]])
```
Will be easier to read if the `new File` is assigned to a variable, and maybe `saveAsHadoopFile` is in two lines.

tdas commented:
There are few nits, but this LGTM even without it. I am merging this. Thanks for this finding and solving this!
